resolvers/dns-resolver/src/resolver.rs

```diff
@@ -0,0 +1,545 @@
+            }
+        }
+    }
+
+    fn poll_a_err<'s, 'c>(
+        _a: &'s mut RentToOwn<'s, AErr>,
+        context: &'c mut RentToOwn<'c, ResolverContext>,
+    ) -> Poll<AfterAErr, ResolverError> {
+        info!(context.log, "srv_err: backends: {:?}", context.backends);
+        transition!(ATry)
+    }
+
+    // process emits events to cueball for each backend
+    fn poll_process<'s, 'c>(
+        _process: &'s mut RentToOwn<'s, Process>,
+        context: &'c mut RentToOwn<'c, ResolverContext>,
+    ) -> Poll<AfterProcess, ResolverError> {
+        let mut new_backends: HashMap<String, backend::Backend> =
+            HashMap::new();
+
+        for srv in context.srvs.iter() {
+            for ip in srv.addresses_v4.iter() {
+                let b = backend::Backend::new(&ip, srv.port);
+                new_backends.entry(b.name.to_string()).or_insert(b);
+            }
+        }
+
+        if new_backends.keys().len() == 0 {
+            info!(
+                context.log,
+                "found no DNS records for {}.{}, next service: {:?}",
+                context.service,
+                context.domain,
+                context.next_service
+            );
+            context.srvs.clear();
+            transition!(Sleep)
+        }
+
+        let mut added: HashMap<String, backend::Backend> = HashMap::new();
+        let mut removed: HashMap<String, backend::Backend> = HashMap::new();
+
+        context.backends.iter().for_each(|(k, v)| {
+            if !new_backends.contains_key(k) {
+                removed.insert(k.to_string(), v.clone());
+            }
+        });
+
+        new_backends.iter().for_each(|(k, v)| {
+            if !context.backends.contains_key(k) {
+                added.insert(k.to_string(), v.clone());
+            }
+        });
+
+        send_updates(added, context.pool_tx.clone(), context.log.clone());
+        send_updates(removed, context.pool_tx.clone(), context.log.clone());
+
+        context.srvs.clear();
+
+        context.backends = new_backends;
+
+        transition!(Sleep)
+    }
+
+    fn poll_sleep<'s, 'c>(
+        _sleep: &'s mut RentToOwn<'s, Sleep>,
+        context: &'c mut RentToOwn<'c, ResolverContext>,
+    ) -> Poll<AfterSleep, ResolverError> {
+        let now = Utc::now().naive_utc();
+
+        let min_delay =
+            context.next_service.unwrap().signed_duration_since(now);
+
+        info!(context.log, "sleeping for {}", min_delay);
+        thread::sleep(min_delay.to_std().unwrap());
+        transition!(Srv)
+    }
+}
+
+fn resolve_srv_records(
+    name: &str,
+    context: &mut ResolverContext,
+) -> Result<(), ResolverError> {
+    for resolver in context.resolvers.iter_mut() {
+        let srv_resp = match resolver.query_srv(&name, &context.log) {
+            Ok(r) => r,
+            Err(e) => {
+                error!(
+                    context.log,
+                    "failed to resolver srv: {}, trying next resolver",
```

Comment:
Nit: should "resolver" be "resolve" here?

Reply:
Fixed, good catch!